pkg/apis/core/validation/validation.go

```diff
@@ -4154,12 +4159,19 @@ func ValidateServiceExternalTrafficFieldsCombination(service *core.Service) fiel
 func ValidateServiceUpdate(service, oldService *core.Service) field.ErrorList {
 	allErrs := ValidateObjectMetaUpdate(&service.ObjectMeta, &oldService.ObjectMeta, field.NewPath("metadata"))
 
-	// ClusterIP should be immutable for services using it (every type other than ExternalName)
+	// ClusterIP and IPFamily should be immutable for services using it (every type other than ExternalName)
 	// which do not have ClusterIP assigned yet (empty string value)
 	if service.Spec.Type != core.ServiceTypeExternalName {
 		if oldService.Spec.Type != core.ServiceTypeExternalName && oldService.Spec.ClusterIP != "" {
 			allErrs = append(allErrs, ValidateImmutableField(service.Spec.ClusterIP, oldService.Spec.ClusterIP, field.NewPath("spec", "clusterIP"))...)
 		}
+		// notes:
+		// we drop the IPFamily field when the Dualstack gate is off.
```

Comment:
This is thought-provoking.

When the gate is off, we will drop the field, so it will be stored as nil.

Once the gate is turned on, we have a bunch of Services that have nil values.

We only set the default on creation.  We've got no real mechanism to back-populate existing services.

Ergo - we HAVE TO allow nil, which means any consumer of this field needs to know how to interpret nil.  Probably "parse the cluster IP" is the best they can do.

Given that, should we even bother with a default?  Maybe the `ipFamily` field is only used for user-defined preference and nil means "whatever the cluster uses" ?

Reply:
to close this line of thought and for future prosperity. 
Typical situation (adding new field to stable api):
1. we nil/blank/remove value on the field (create/update) if the gate is off.
2. apply a static default if the gate is on. 

for IPFamily, because it is not a `static default` we:
1. keep on nil/blanking/removing as usual.
2. set a dynamic default if the gate is on, because of that the default is dependent on other field value in case of `Service.Spec.IPFamily` it is dependent on `Service.Spec.ClusterIP`.